### PR TITLE
Sabakan TLS step1

### PR DESCRIPTION
**Pull request by YZ775**

### CODE REVIEW

File: constants.go [hunk 1]
```diff
@@ -139,6 +139,9 @@ var (
 	ServerCertFile = filepath.Join(NecoDir, "server.crt")
 	ServerKeyFile  = filepath.Join(NecoDir, "server.key")
 
+	SabakanTLSCertFile = filepath.Join(SabakanDir, "sabakan-tls.crt")
+	SabakanTLSKeyFile  = filepath.Join(SabakanDir, "sabakan-tls.key")
```
Comment (masa213f) [inline constants.go:143]: Please make it clear that they are server certificates.

```suggestion
	SabakanServerCertFile = filepath.Join(SabakanDir, "sabakan-server.crt")
	SabakanServerKeyFile  = filepath.Join(SabakanDir, "sabakan-server.key")
```

File: pkg/neco/cmd/init_local.go [hunk 2]
```diff
@@ -72,7 +84,15 @@ new a application NAME.`,
...
-				err = neco.StartService(ctx, neco.SabakanService)
+				var isActive bool
+				isActive, err = neco.IsActiveService(ctx, neco.SabakanService)
+				if err != nil {
+					return err
+				}
+				if isActive {
+					return neco.RestartService(ctx, neco.SabakanService)
+				}
+				return neco.StartService(ctx, neco.SabakanService)
```
Comment (masa213f) [inline pkg/neco/cmd/init_local.go:95]: Do we need to restart the sabakan service at this time?
When we will run `neco-init sabakan`, the new sabakan will not be released yet, won't it?

File: pkg/neco/cmd/init_local.go [hunk 3]
```diff
@@ -95,6 +115,10 @@ func init() {
 }
 
 func issueCerts(ctx context.Context, vc *api.Client, commonName, cert, key string) error {
+	_, err := os.Stat(cert)
+	if err == nil {
+		return os.ErrExist
+	}
```
Comment (masa213f) [inline pkg/neco/cmd/init_local.go:121]: Why did you add this check?

File: pkg/neco/cmd/init_local.go [hunk 4]
```diff
@@ -109,3 +133,32 @@ func issueCerts(ctx context.Context, vc *api.Client, commonName, cert, key strin
 	return neco.WriteFile(key, secret.Data["private_key"].(string))
 
 }
+
+func issueSabakanTLSCerts(ctx context.Context, vc *api.Client, cert, key string) error {
```
Comment (masa213f) [inline pkg/neco/cmd/init_local.go:137]: Please make it clear that this is for server certificates.

```suggestion
func issueSabakanServerCerts(ctx context.Context, vc *api.Client, cert, key string) error {
```

File: pkg/neco/cmd/init_data.go [hunk 5]
```diff
@@ -74,6 +75,15 @@ func initData(ctx context.Context, st storage.Storage) error {
 		return cke.UpdateResources(ctx, st)
 	}
 
+	if uploadCACertOnly {
+		sabakanClient, err := client.NewClient(neco.SabakanLocalEndpoint, httpClient.Client)
+		if err != nil {
+			return err
+		}
+		_, err = sabakanClient.AssetsUpload(ctx, "sabakan-tls-ca.crt", neco.ServerCAFile, nil)
```
Comment (masa213f) [inline pkg/neco/cmd/init_data.go:83]: Please create a function in `progs/sabakan` for this logic.

**Review by masa213f — CHANGES_REQUESTED** [review]

File: pkg/neco/cmd/init_local.go (hunk 2 above)
Comment (YZ775) [reply pkg/neco/cmd/init_local.go:95]: I wrote this code before I came up with the idea of splitting deploy operations to 2 step, so I expected that update of  sabakan and neco command would be same time.
I'll remove this restart logic.

File: pkg/neco/cmd/init_local.go (hunk 3 above)
Comment (YZ775) [reply pkg/neco/cmd/init_local.go:121]: I want to make `neco-init sabakan` command to be idempotent. so I need to send the existence of the cert file.
This check returns `ErrExist` when the cert file are already exists in specified directry.

Comment (masa213f) [reply pkg/neco/cmd/init_local.go:120]: Are there any side effects of issuing a new cert?

I assume that the `neco init-local` is executed only once during initializing the boot server. So, it is not necessary to be idempotent.

If the side effects of issuing a new cert are negligible, we can overwrite it.

File: pkg/neco/cmd/init_local.go [hunk 6]
```diff
@@ -72,7 +84,14 @@ new a application NAME.`,
 			case "etcdpasswd":
 				err = neco.StartService(ctx, neco.EtcdpasswdService)
 			case "sabakan":
-				err = neco.StartService(ctx, neco.SabakanService)
+				var isActive bool
```
Comment (masa213f) [inline pkg/neco/cmd/init_local.go:87]: Do you avoid the error by turning off the sabakan service before the operation?
I don't want to leave code that executes only once for updating the sabakan.

**Review by masa213f — CHANGES_REQUESTED** [review]

File: pkg/neco/cmd/init_local.go [hunk 7]
```diff
@@ -52,15 +54,37 @@ new a application NAME.`,
 			var err error
 			switch initLocalParams.name {
 			case "etcdpasswd":
-				err = issueCerts(ctx, vc, "etcdpasswd", neco.EtcdpasswdCertFile, neco.EtcdpasswdKeyFile)
+				err = issueEtcdCerts(ctx, vc, "etcdpasswd", neco.EtcdpasswdCertFile, neco.EtcdpasswdKeyFile)
+				if err != nil {
+					if err != os.ErrExist {
+						return err
+					}
+					fmt.Printf("etcdpasswd certificate already exists in %s.  skipping issue cert\n", neco.SabakanEtcdCertFile)
```
Comment (masa213f) [inline pkg/neco/cmd/init_local.go:62]:
```suggestion
					fmt.Printf("etcdpasswd certificate already exists in %s.  skipping issue cert\n", neco.EtcdpasswdCertFile)
```

File: pkg/neco/cmd/init_local.go [hunk 8]
```diff
@@ -52,15 +54,37 @@ new a application NAME.`,
...
 				}
 				err = sabakan.InitLocal(ctx, vc)
 			case "cke":
-				err = issueCerts(ctx, vc, "cke", neco.CKECertFile, neco.CKEKeyFile)
+				err = issueEtcdCerts(ctx, vc, "cke", neco.CKECertFile, neco.CKEKeyFile)
+				if err != nil {
+					if err != os.ErrExist {
+						return err
+					}
+					fmt.Printf("cke certificate already exists in %s.  skipping issue cert\n", neco.SabakanEtcdCertFile)
```
Comment (masa213f) [inline pkg/neco/cmd/init_local.go:86]:
```suggestion
					fmt.Printf("cke certificate already exists in %s.  skipping issue cert\n", neco.CKECertFile)
```

File: pkg/neco/cmd/init_local.go [hunk 9]
```diff
@@ -52,15 +54,37 @@ new a application NAME.`,
...
+				}
 			case "sabakan":
-				err = issueCerts(ctx, vc, "sabakan", neco.SabakanCertFile, neco.SabakanKeyFile)
+				err = issueEtcdCerts(ctx, vc, "sabakan", neco.SabakanEtcdCertFile, neco.SabakanEtcdKeyFile)
 				if err != nil {
-					return err
+					if err != os.ErrExist {
+						return err
+					}
+					fmt.Printf("sabakan certificate already exists in %s.  skipping issue cert\n", neco.SabakanEtcdCertFile)
```
Comment (masa213f) [inline pkg/neco/cmd/init_local.go:70]: Why don't you use `log.Info()`?

**Review by masa213f — CHANGES_REQUESTED** [review]

**Review by masa213f — APPROVED** [review]
LGTM

**Review by masa213f — APPROVED** [review]

**Pull request merged**